### PR TITLE
[6X] Enable uuid-ossp extension by default

### PR DESCRIPTION
**Pull request by liuxueyang**
Users want to enable the `contrib/uuid-ossp` extension by default. This
extension requires the uuid library.

Backport https://github.com/greenplum-db/gpdb/pull/14172 to 6X_STABLE. Postgresql 9.4 does not support
`PARALLEL SAFE` option in CREATE/ALTER function.

Co-authored-by: Xing Guo <admin@higuoxing.com>
Co-authored-by: Chen Mulong <chenmulong@gmail.com>

### CODE REVIEW

**Comment by yihong0618** [comment]
CI failed 
```
checking for uuid_generate in -luuid... no
configure: error: library 'uuid' is required for E2FS UUID
```

**Comment by beeender** [comment]
If `libuuid` is not installed, what will happen? Only the extension cannot be used or the `postgres` cannot be started? what is the output of `ldd <path_to_gpdb>/bin/postgres`?

**Comment by liuxueyang** [comment]
@beeender 
> If `libuuid` is not installed, what will happen? Only the extension cannot be used or the `postgres` cannot be started? what is the output of `ldd <path_to_gpdb>/bin/postgres`?

On centos7, if `libuuid-devel` package is not installed, and the gpdb binary is installed, then this extension works as usual. `make installcheck` passed. But `configure` with `with-uuid` enabled will fail. I think the reason is that the uuid library is part of the Linux operating system. The `libuuid-devel` package only did something or create some symlinks of uuid.

I guess what you want is the output of `ldd uuid-ossp.so`:

```
gpadmin in localhost in gpdb/lib/postgresql
❯ ldd uuid-ossp.so
        linux-vdso.so.1 =>  (0x00007ffe4bd92000)
        libuuid.so.1 => /lib64/libuuid.so.1 (0x00007f8d69dff000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f8d69a31000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f8d6a004000)
```

On my dev machine, `.so` files about libuuid:

```
gpadmin in localhost in gpdb/lib/postgresql
❯ ls -l /lib64/libuuid.*
lrwxrwxrwx  1 root root    16 9月  29 15:54 /lib64/libuuid.so -> libuuid.so.1.3.0
lrwxrwxrwx. 1 root root    16 1月  11 2022 /lib64/libuuid.so.1 -> libuuid.so.1.3.0
-rwxr-xr-x. 1 root root 20064 2月   3 2021 /lib64/libuuid.so.1.3.0
```

After `yum remove libuuid-devel`

```
gpadmin in localhost in gpdb/lib/postgresql took 6s
❯ ls -l /lib64/libuuid.*
lrwxrwxrwx. 1 root root    16 1月  11 2022 /lib64/libuuid.so.1 -> libuuid.so.1.3.0
-rwxr-xr-x. 1 root root 20064 2月   3 2021 /lib64/libuuid.so.1.3.0
```

Only the symlink file `libuuid.so` is removed. So I guess the symlink file `libuuid.so` is created by `libuuid-devel` package. And the other two files are provided by the `libuuid` package.

**Comment by beeender** [comment]
> I guess what you want is the output of ldd uuid-ossp.so:

Actually, I wanted the result of `ldd postgres`. I want to make sure if the user doesn't load the extension, gpdb can be started without `libuuid` as before. Otherwise, we need to change the gpdb document.

**Comment by liuxueyang** [comment]
> > I guess what you want is the output of ldd uuid-ossp.so:
> 
> Actually, I wanted the result of `ldd postgres`. I want to make sure if the user doesn't load the extension, gpdb can be started without `libuuid` as before. Otherwise, we need to change the gpdb document.

This is the output of `ldd postgres`

```
gpadmin in localhost in gpdb/contrib/ltree on  6X_STABLE
❯ ldd ~/local/gpdb/bin/postgres
        linux-vdso.so.1 =>  (0x00007ffdc1770000)
        libnuma.so.1 => /lib64/libnuma.so.1 (0x00007f12f4149000)
        libbz2.so.1 => /lib64/libbz2.so.1 (0x00007f12f3f39000)
        libxml2.so.2 => /lib64/libxml2.so.2 (0x00007f12f3bcf000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f12f39c7000)
        libzstd.so.1 => /lib64/libzstd.so.1 (0x00007f12f3714000)
        libz.so.1 => /usr/local/lib/libz.so.1 (0x00007f12f4527000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f12f3510000)
        libcurl.so.4 => /lib64/libcurl.so.4 (0x00007f12f32a6000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f12f308a000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f12f2d88000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f12f2b72000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f12f27a4000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f12f4355000)
        liblzma.so.5 => /lib64/liblzma.so.5 (0x00007f12f257e000)
        libidn.so.11 => /lib64/libidn.so.11 (0x00007f12f234b000)
        libssh2.so.1 => /lib64/libssh2.so.1 (0x00007f12f211e000)
        libssl3.so => /lib64/libssl3.so (0x00007f12f1ebb000)
        libsmime3.so => /lib64/libsmime3.so (0x00007f12f1c93000)
        libnss3.so => /lib64/libnss3.so (0x00007f12f195a000)
        libnssutil3.so => /lib64/libnssutil3.so (0x00007f12f172a000)
        libplds4.so => /lib64/libplds4.so (0x00007f12f1526000)
        libplc4.so => /lib64/libplc4.so (0x00007f12f1321000)
        libnspr4.so => /lib64/libnspr4.so (0x00007f12f10e3000)
        libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007f12f0e96000)
        libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007f12f0bad000)
        libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007f12f097a000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f12f0776000)
        liblber-2.4.so.2 => /lib64/liblber-2.4.so.2 (0x00007f12f0567000)
        libldap-2.4.so.2 => /lib64/libldap-2.4.so.2 (0x00007f12f0312000)
        libssl.so.10 => /lib64/libssl.so.10 (0x00007f12f00a0000)
        libcrypto.so.10 => /lib64/libcrypto.so.10 (0x00007f12efc3d000)
        libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007f12efa2d000)
        libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007f12ef829000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f12ef60f000)
        libsasl2.so.3 => /lib64/libsasl2.so.3 (0x00007f12ef3f2000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f12ef1cb000)
        libcrypt.so.1 => /lib64/libcrypt.so.1 (0x00007f12eef94000)
        libpcre.so.1 => /lib64/libpcre.so.1 (0x00007f12eed32000)
        libfreebl3.so => /lib64/libfreebl3.so (0x00007f12eeb2f000)
```

File: contrib/uuid-ossp/Makefile [hunk 1]
```diff
@@ -1,12 +1,14 @@
 # contrib/uuid-ossp/Makefile
 
 MODULE_big = uuid-ossp
-OBJS = uuid-ossp.o $(UUID_EXTRA_OBJS)
+OBJS = uuid-ossp.o $(UUID_EXTRA_OBJS) $(WIN32RES)
```
Comment (beeender) [inline contrib/uuid-ossp/Makefile:4]: What is this `$(WIN32RES)`? Do we need this ?

**Comment by beeender** [comment]
> Keep version 1.0 on stable branch

Hmmm, I actually think we should release 1.1 directly, to save the upgrade pain in the future.

What is your reason to roll back to 1.0?

**Comment by liuxueyang** [comment]
> > Keep version 1.0 on stable branch
> 
> Hmmm, I actually think we should release 1.1 directly, to save the upgrade pain in the future.
> 
> What is your reason to roll back to 1.0?

Actually, those changes I migrated from main to 6X_STABLE do not contain the newer changes in version 1.1. I think it's not suitable to call it version 1.1

File: contrib/uuid-ossp/Makefile (hunk 1 above)
Comment (liuxueyang) [reply contrib/uuid-ossp/Makefile:4]: I pick this from the main branch. Maybe it's used by win32 binaries.

**Review by beeender — APPROVED** [review]

**Comment by liuxueyang** [comment]
> CI failed
> 
> ```
> checking for uuid_generate in -luuid... no
> configure: error: library 'uuid' is required for E2FS UUID
> ```

The CI succeeds now. :P

**Review by Sasasu — APPROVED** [review]
there are two files linked to `libuuid`
 - ./bin/gpfdist
 - ./lib/postgresql/uuid-ossp.so (this one)

**Review by kainwen — APPROVED** [review]

File: contrib/uuid-ossp/uuid-ossp.c [hunk 2]
```diff
@@ -505,13 +505,13 @@ Datum
 uuid_generate_v3(PG_FUNCTION_ARGS)
 {
 	pg_uuid_t  *ns = PG_GETARG_UUID_P(0);
-	text	   *name = PG_GETARG_TEXT_PP(1);
```
Comment (liuxueyang) [inline contrib/uuid-ossp/uuid-ossp.c]: `PG_GETARG_TEXT_PP` and `PG_GETARG_TEXT_P` are equal for the text type argument. In this case, I think it's safe to stay with postgres_9.4 though the new version of Postgres recommended the `PP` version.

The corresponding commit of Postgres are: 
1. [9d7726c2ba06b932f791f2d0cc5acf73cc0b4dca](https://www.postgresql.org/message-id/20161025145724.GA1547663%40tornado.leadboat.com)
2. [3a0d473192b2045cbaf997df8437e7762d34f3ba](https://www.postgresql.org/message-id/E1cnD7M-00034N-U1@gemulon.postgresql.org)
3. [3e23b68dac006e8deb0afa327e855258df8de064](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=3e23b68dac006e8deb0afa327e855258df8de064)

**Pull request merged**